### PR TITLE
Try to fix filtering of feature files in intellij support code

### LINKED ISSUE #31: Support other build systems in zuchini-intellij-support
Currently `ZuchiniMain` of zuchini-intellij-support assumes a Maven style default build folder structure (see [here](https://github.com/jhorstmann/zuchini/blob/master/zuchini-intellij-support/src/main/java/org/zuchini/intellij/ZuchiniMain.java#L61). This does not work with other build systems like Gradle.

Default paths on Gradle look like:

```
~/my-project/out/test/resources/hello.feature
```

which does not match

```
~/my-project/src/test/resources/hello.feature
```
and the replacement String `/target/test-classes/` is not effective. Mind that the source set name `test` is just a convention and might be different for other projects.

### PR DESCRIPTION
**Pull request by jhorstmann**
Fixes #31

### CODE REVIEW

**Comment by lukasniemeier-zalando** [comment]
I have tested this change against a real-world gradle project and a (hopefully) minimal: 

**Good**: the correct scenarios do run now when started from IntelliJ IDEA.
**Bad**: after all (filtered) scenarios are executed the test process is stuck and never ends.

What I can see is that the main entry-point of `ZuchiniMain::main` does terminate, but something in my IDEA is still waiting on something after that.

**Comment by jhorstmann** [comment]
@lukasniemeier-zalando Interesting, this should really behave the same as running the JUnit Runner directly from IntelliJ. Could you get a thread dump to see what is waiting? Another option for investigateion would be to run `ZuchiniMain` from the command line, with the parameters taken from the IntelliJ run configuration.

**Comment by lukasniemeier-zalando** [comment]
Following is the thread dump of the not-so-minimal example, I will try the other thing later:

```
Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.121-b13 mixed mode):

"DestroyJavaVM" #34 prio=5 os_prio=0 tid=0x00007f1c10017800 nid=0x5a2e waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"http-nio-auto-1-44259-AsyncTimeout" #32 daemon prio=5 os_prio=0 tid=0x00007f1c11891000 nid=0x5ae5 waiting on condition [0x00007f1bbfefd000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.apache.coyote.AbstractProtocol$AsyncTimeout.run(AbstractProtocol.java:1211)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-Acceptor-0" #31 daemon prio=5 os_prio=0 tid=0x00007f1c1188f800 nid=0x5ae4 runnable [0x00007f1bc419b000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method)
	at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:422)
	at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:250)
	- locked <0x0000000763d0ce28> (a java.lang.Object)
	at org.apache.tomcat.util.net.NioEndpoint$Acceptor.run(NioEndpoint.java:455)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-ClientPoller-1" #30 daemon prio=5 os_prio=0 tid=0x00007f1c1188d000 nid=0x5ae3 runnable [0x00007f1bc429c000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x0000000763eaff70> (a sun.nio.ch.Util$3)
	- locked <0x0000000763eaff60> (a java.util.Collections$UnmodifiableSet)
	- locked <0x0000000763eafe48> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at org.apache.tomcat.util.net.NioEndpoint$Poller.run(NioEndpoint.java:793)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-ClientPoller-0" #29 daemon prio=5 os_prio=0 tid=0x00007f1c11872800 nid=0x5ae2 runnable [0x00007f1bc439d000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x0000000763e9f580> (a sun.nio.ch.Util$3)
	- locked <0x0000000763e9f570> (a java.util.Collections$UnmodifiableSet)
	- locked <0x0000000763e9f458> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at org.apache.tomcat.util.net.NioEndpoint$Poller.run(NioEndpoint.java:793)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-10" #28 daemon prio=5 os_prio=0 tid=0x00007f1c11856800 nid=0x5ae1 waiting on condition [0x00007f1bc449e000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-9" #27 daemon prio=5 os_prio=0 tid=0x00007f1c11854800 nid=0x5ae0 waiting on condition [0x00007f1bc459f000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-8" #26 daemon prio=5 os_prio=0 tid=0x00007f1c11852800 nid=0x5adf waiting on condition [0x00007f1bc46a0000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-7" #25 daemon prio=5 os_prio=0 tid=0x00007f1c11851000 nid=0x5ade waiting on condition [0x00007f1bc47a1000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-6" #24 daemon prio=5 os_prio=0 tid=0x00007f1c1184f000 nid=0x5add waiting on condition [0x00007f1bc48a2000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-5" #23 daemon prio=5 os_prio=0 tid=0x00007f1c1184d000 nid=0x5adc waiting on condition [0x00007f1bc49a3000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-4" #22 daemon prio=5 os_prio=0 tid=0x00007f1c1184b000 nid=0x5adb waiting on condition [0x00007f1bc4aa4000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-3" #21 daemon prio=5 os_prio=0 tid=0x00007f1c11849000 nid=0x5ada waiting on condition [0x00007f1bc4ba5000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-2" #20 daemon prio=5 os_prio=0 tid=0x00007f1c11847800 nid=0x5ad9 waiting on condition [0x00007f1bc4ca6000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"http-nio-auto-1-exec-1" #19 daemon prio=5 os_prio=0 tid=0x00007f1c11846000 nid=0x5ad8 waiting on condition [0x00007f1bc4da7000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000763d81e70> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:103)
	at org.apache.tomcat.util.threads.TaskQueue.take(TaskQueue.java:31)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)

"NioBlockingSelector.BlockPoller-1" #18 daemon prio=5 os_prio=0 tid=0x00007f1c11843000 nid=0x5ad7 runnable [0x00007f1bc4ea8000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x0000000763d1abb8> (a sun.nio.ch.Util$3)
	- locked <0x0000000763d1a320> (a java.util.Collections$UnmodifiableSet)
	- locked <0x0000000763d182a0> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at org.apache.tomcat.util.net.NioBlockingSelector$BlockPoller.run(NioBlockingSelector.java:339)

"container-0" #17 prio=5 os_prio=0 tid=0x00007f1c10dcf000 nid=0x5ab6 waiting on condition [0x00007f1bc5b64000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.apache.catalina.core.StandardServer.await(StandardServer.java:427)
	at org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer$1.run(TomcatEmbeddedServletContainer.java:177)

"ContainerBackgroundProcessor[StandardEngine[Tomcat]]" #16 daemon prio=5 os_prio=0 tid=0x00007f1c10dce800 nid=0x5ab5 waiting on condition [0x00007f1bc672b000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.run(ContainerBase.java:1355)
	at java.lang.Thread.run(Thread.java:745)

"Service Thread" #9 daemon prio=9 os_prio=0 tid=0x00007f1c10851800 nid=0x5a49 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread2" #8 daemon prio=9 os_prio=0 tid=0x00007f1c1084e800 nid=0x5a48 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #7 daemon prio=9 os_prio=0 tid=0x00007f1c1084a800 nid=0x5a47 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #6 daemon prio=9 os_prio=0 tid=0x00007f1c10849800 nid=0x5a46 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Monitor Ctrl-Break" #5 daemon prio=5 os_prio=0 tid=0x00007f1c1084c800 nid=0x5a45 runnable [0x00007f1bdf520000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	- locked <0x000000068f81a350> (a java.io.InputStreamReader)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	- locked <0x000000068f81a350> (a java.io.InputStreamReader)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at com.intellij.rt.execution.application.AppMainV2$1.run(AppMainV2.java:64)

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 tid=0x00007f1c1018c800 nid=0x5a44 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x00007f1c10159000 nid=0x5a3d in Object.wait() [0x00007f1bdfffe000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000068f81a500> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x000000068f81a500> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

"Reference Handler" #2 daemon prio=10 os_prio=0 tid=0x00007f1c10154800 nid=0x5a3c in Object.wait() [0x00007f1bf410f000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000068f81a540> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x000000068f81a540> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

"VM Thread" os_prio=0 tid=0x00007f1c1014c800 nid=0x5a3b runnable 

"GC task thread#0 (ParallelGC)" os_prio=0 tid=0x00007f1c1002d000 nid=0x5a34 runnable 

"GC task thread#1 (ParallelGC)" os_prio=0 tid=0x00007f1c1002f000 nid=0x5a35 runnable 

"GC task thread#2 (ParallelGC)" os_prio=0 tid=0x00007f1c10030800 nid=0x5a36 runnable 

"GC task thread#3 (ParallelGC)" os_prio=0 tid=0x00007f1c10032800 nid=0x5a37 runnable 

"VM Periodic Task Thread" os_prio=0 tid=0x00007f1c10857000 nid=0x5a4a waiting on condition 

JNI global references: 601

Heap
 PSYoungGen      total 280064K, used 273093K [0x000000075a780000, 0x0000000770400000, 0x00000007c0000000)
  eden space 265728K, 98% used [0x000000075a780000,0x000000076a68d790,0x000000076ab00000)
  from space 14336K, 83% used [0x000000076ab00000,0x000000076b6a4058,0x000000076b900000)
  to   space 15360K, 0% used [0x000000076f500000,0x000000076f500000,0x0000000770400000)
 ParOldGen       total 202752K, used 19475K [0x000000068f600000, 0x000000069bc00000, 0x000000075a780000)
  object space 202752K, 9% used [0x000000068f600000,0x0000000690904d28,0x000000069bc00000)
 Metaspace       used 50593K, capacity 51622K, committed 51840K, reserved 1093632K
  class space    used 6499K, capacity 6739K, committed 6784K, reserved 1048576K
```

**Comment by whiskeysierra** [comment]
This is one of two *non-daemon* threads:

```
"container-0" #17 prio=5 os_prio=0 tid=0x00007f1c10dcf000 nid=0x5ab6 waiting on condition [0x00007f1bc5b64000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.apache.catalina.core.StandardServer.await(StandardServer.java:427)
	at org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer$1.run(TomcatEmbeddedServletContainer.java:177)
```

The other is called `DestroyJavaVM`, so I'd assume it's not the cause.

**Comment by jhorstmann** [comment]
I can reproduce this without gradle, by adding the `zuchini-intellij-support` dependency to the `zuchini-examples-mockmvc` project and changing the test from `WebEnvironment.MOCK` to `WebEnvironment.RANDOM_PORT`.

**Comment by whiskeysierra** [comment]
> changing the test from WebEnvironment.MOCK to WebEnvironment.RANDOM_PORT.

Which probably changes the test from `MockMvc` to `EmbeddedTomcatServletContainer`.

**Comment by jhorstmann** [comment]
It looks like Intellij does an explicit `System.exit` when running JUnit tests: https://github.com/JetBrains/intellij-community/blob/7fa96faf725f9ee605444bacb794c21906a3e699/plugins/junit_rt/src/com/intellij/rt/execution/junit/JUnitStarter.java#L71

**Comment by jhorstmann** [comment]
And cucumber-jvm does also call `System.exit` explicitly: https://github.com/cucumber/cucumber-jvm/blob/f7000b2545bbf433377ce4730313af8264d23549/core/src/main/java/cucumber/api/cli/Main.java#L19

I'll add this to `ZuchiniMain`.

File: zuchini-intellij-support/src/main/java/org/zuchini/intellij/ZuchiniMain.java [hunk 1]
```diff
@@ -39,10 +40,18 @@ public static void main(String[] args) throws Throwable {
...
 
-        run(cl, featureFiles, glue, name);
+        // Exit the jvm to stop any running non-daemon threads, for example embedded servlet containers
+        // Use same exit codes as IntelliJ JUnit4IdeaTestRunner
+        // See https://github.com/JetBrains/intellij-community/blob/7fa96faf725f9ee605444bacb794c21906a3e699/plugins/junit_rt/src/com/intellij/junit4/JUnit4IdeaTestRunner.java#L69
+        try {
+            Result result = run(cl, featureFiles, glue, name);
+            System.exit(result.wasSuccessful() ? 0 : -1);
+        } catch (Throwable throwable) {
+            System.exit(-2);
```
Comment (whiskeysierra) [inline zuchini-intellij-support/src/main/java/org/zuchini/intellij/ZuchiniMain.java]: Print stacktrace?

Comment (whiskeysierra) [reply zuchini-intellij-support/src/main/java/org/zuchini/intellij/ZuchiniMain.java]: I believe you get the catch clause for free, i.e. you don't need to handle the exception.

Comment (whiskeysierra) [reply zuchini-intellij-support/src/main/java/org/zuchini/intellij/ZuchiniMain.java]: I can't remember which exit status you'd get, but it's not 0.

**Comment by lukasniemeier-zalando** [comment]
image

It is working better now, but something is still off. After all scenarios are executed and reported as successful, IDEA complains that it was unable to *start* any test.

**Comment by jhorstmann** [comment]
@lukasniemeier-zalando I changed the test reporting a little bit, can you try again?

This reporting is a total hack btw, but IntelliJ does the same for their JUnit integration: https://github.com/JetBrains/intellij-community/blob/7fa96faf725f9ee605444bacb794c21906a3e699/plugins/junit_rt/src/com/intellij/junit4/JUnit4TestListener.java

**Comment by lukasniemeier-zalando** [comment]
I've just checked and my IDE is behaving well now :1st_place_medal: I've tested both with and without `reportIndividualSteps`. Looks fine now (functionally :smirk:)

**Review by lukasniemeier-zalando — APPROVED** [review]

**Pull request merged**